alacritty_terminal/src/renderer/rects.rs

```diff
@@ -35,69 +35,66 @@ impl<T> Rect<T> {
-            Flags::STRIKEOUT => (metrics.strikeout_position, metrics.strikeout_thickness),
-            _ => unimplemented!("Invalid flag for cell line drawing specified"),
-        };
-
-        // Make sure lines are always visible
-        height = height.max(1.);
-
-        let cell_bottom = (cell.line.0 as f32 + 1.) * size.cell_height;
-        let baseline = cell_bottom + metrics.descent;
-
-        let mut y = baseline - position - height / 2.;
-        let max_y = cell_bottom - height;
-        if y > max_y {
-            y = max_y;
-        }
-
-        let rect = Rect::new(cell_x + size.padding_x, y + size.padding_y, size.cell_width, height);
-
-        Self { start: cell.into(), end: cell.into(), color: cell.fg, rect }
-    }
+/// Lines for underline and strikeout.
+#[derive(Default)]
+pub struct Lines {
+    inner: HashMap<Flags, Vec<Line>>,
+}
 
-    fn update_end(&mut self, end: Point, size: &SizeInfo) {
-        self.rect.width = (end.col + 1 - self.start.col).0 as f32 * size.cell_width;
-        self.end.col = end.col;
+fn create_rect(
+    line: &Line,
+    flag: Flags,
+    metrics: &Metrics,
+    size: &SizeInfo,
+    color: Rgb,
+) -> (Rect<f32>, Rgb) {
+    let start_x = line.start.col.0 as f32 * size.cell_width;
+    let end_x = (line.end.col.0 + 1) as f32 * size.cell_width;
+    let width = end_x - start_x;
+
+    let (position, mut height) = match flag {
+        Flags::UNDERLINE => (metrics.underline_position, metrics.underline_thickness),
+        Flags::STRIKEOUT => (metrics.strikeout_position, metrics.strikeout_thickness),
+        _ => unimplemented!("Invalid flag for cell line drawing specified"),
+    };
+
+    // Make sure lines are always visible
+    height = height.max(1.);
+
+    let line_bottom = (line.start.line.0 as f32 + 1.) * size.cell_height;
+    let baseline = line_bottom + metrics.descent;
+
+    let mut y = baseline - position - height / 2.;
+    let max_y = line_bottom - height;
+    if y > max_y {
+        y = max_y;
     }
-}
 
-/// Rects for underline, strikeout and more.
-#[derive(Default)]
-pub struct Rects {
-    inner: HashMap<Flags, Vec<Line>>,
+    (Rect::new(start_x + size.padding_x, y + size.padding_y, width, height), color)
 }
 
-impl Rects {
+impl Lines {
     pub fn new() -> Self {
         Self::default()
     }
 
-    /// Convert the stored rects to rectangles for the renderer.
-    pub fn rects(&self) -> Vec<(Rect<f32>, Rgb)> {
-        self.inner
-            .iter()
-            .map(|(_, lines)| lines)
-            .flatten()
-            .map(|line| (line.rect, line.color))
-            .collect()
+    pub fn into_rects(self, metrics: &Metrics, size: &SizeInfo) -> Vec<(Rect<f32>, Rgb)> {
+        let mut rects = Vec::with_capacity(self.inner.capacity());
+        for (flag, lines) in self.inner {
+            for line in lines {
+                rects.push(create_rect(&line, flag, &metrics, &size, line.color));
+            }
+        }
+        rects
     }
```

Comment:
Can this not be done with an iterator? That would save us from having to first define a vector and then fill it bit by bit.

Reply:
I've tried to create a working solution here to expose structure. I'll try to create an iterator here, just It takes me more time, than such for loops.

Reply:
I think `self.inner.drain(..).map(|line| create_rect(...))..collect()` should work.

Reply:
After some time I give up on this one, since things were more complicated with iterators here, since you have 2 loops and HashMap + Vec.

Reply:
You can create a vec from a hashmap using flatten, just take a look at the previous code.

`self.inner.drain(..).map(|(_, lines)| lines).flatten()` should create an iterator over `Line`, then from there my previous comment with the `.map(|line| create_rect(...)).collect()` should be applicable.

Reply:
Yeah, but in my case I need something like Vec<(Flags, Line)> at some point, since I need Flag, which isn't the case in your previous code.

Reply:
`self.inner.drain(..).flatten()` will give you an iterator over `(Flags, Line)`, that should be even easier.

Reply:
`self.inner.drain().flatten()` wouldn't compile since you'll need to implement all that stuff `the trait `std::iter::Iterator` is not implemented for `(term::cell::Flags, std::vec::Vec<renderer::rects::Line>)``

Reply:
I've implemented this one in iterators though